File: orchestrator/apps/templates/workflow/handlers/loop.py
"""
LoopHandler for Workflow Engine.

Executes loop nodes with count/while/foreach modes.
"""

import logging
import time
from typing import Any, Dict, List

from jinja2 import StrictUndefined
from jinja2.sandbox import ImmutableSandboxedEnvironment

from apps.templates.workflow.models import WorkflowExecution, WorkflowNode

from .base import BaseNodeHandler, NodeExecutionMode, NodeExecutionResult

logger = logging.getLogger(__name__)


class LoopHandler(BaseNodeHandler):
    """
    Handler for Loop nodes.

    Loop modes:
        - count: for i in range(N)
        - while: while condition_is_true (sandboxed Jinja2)
        - foreach: for item in collection

    Context updates for each iteration:
        - loop.index: Current iteration index (0-based)
        - loop.first: True if first iteration
        - loop.last: True if last iteration
        - item: Current item (foreach mode only)

    Safety:
        - max_iterations hardcoded limit (10000)
        - Sandboxed Jinja2 for while condition evaluation
        - ImmutableSandboxedEnvironment with StrictUndefined

    Current Implementation (Week 8):
        - Uses placeholder execute_workflow_node task (Week 9)
        - SYNC mode only
    """

    # Safety limit for loop iterations
    MAX_ITERATIONS_HARD_LIMIT = 10000

    def __init__(self):
        """Initialize LoopHandler with sandboxed Jinja2 environment."""
        # Setup sandbox environment for while condition evaluation
        self.env = ImmutableSandboxedEnvironment(
            autoescape=False,
            trim_blocks=True,
            lstrip_blocks=True,
            undefined=StrictUndefined
        )

        # Whitelist safe globals
        self.env.globals.update({
            'len': len,
            'str': str,
            'int': int,
            'float': float,
            'bool': bool,
        })

    def execute(
        self,
        node: WorkflowNode,
        context: Dict[str, Any],
        execution: WorkflowExecution,
        mode: NodeExecutionMode = NodeExecutionMode.SYNC
    ) -> NodeExecutionResult:
        """
        Execute loop node based on mode.

        Args:
            node: WorkflowNode with loop_config
            context: Execution context with variables
            execution: WorkflowExecution for tracking
            mode: Execution mode (currently only SYNC supported)

        Returns:
            NodeExecutionResult with iteration results or error
        """
        start_time = time.time()

        # Create step result for audit
        step_result = self._create_step_result(
            execution=execution,
            node=node,
            input_data={'context_keys': list(context.keys())}
        )

        try:
            # 1. Get loop config
            if not node.loop_config:
                raise ValueError(f"Loop node {node.id} missing loop_config")

            loop_config = node.loop_config
            loop_mode = loop_config.mode
            loop_node_id = loop_config.loop_node_id
            max_iterations = min(loop_config.max_iterations, self.MAX_ITERATIONS_HARD_LIMIT)

            logger.info(
                f"Executing loop node {node.id}",
                extra={
                    'node_id': node.id,
                    'mode': loop_mode,
                    'loop_node_id': loop_node_id,
                    'max_iterations': max_iterations
                }
            )

            # 2. Import Celery task (placeholder in Week 8)
            try:
                from apps.templates.tasks import execute_workflow_node
            except ImportError:
                raise NotImplementedError(
                    "execute_workflow_node task not implemented yet (Week 9)"
                )

            # 3. Execute loop based on mode
            if loop_mode == "count":
                result_data = self._execute_count_loop(
                    loop_config, loop_node_id, context, execution, max_iterations
                )
            elif loop_mode == "while":
                result_data = self._execute_while_loop(
                    loop_config, loop_node_id, context, execution, max_iterations
                )
            elif loop_mode == "foreach":
                result_data = self._execute_foreach_loop(
                    loop_config, loop_node_id, context, execution, max_iterations
                )
            else:
                raise ValueError(f"Unknown loop mode: {loop_mode}")

            duration = time.time() - start_time

            # 4. Return success result
            result = NodeExecutionResult(
                success=True,
                output=result_data,
                error=None,
                mode=NodeExecutionMode.SYNC,
                duration_seconds=duration
            )

            self._update_step_result(step_result, result)

            logger.info(
                f"Loop node {node.id} completed successfully",
                extra={
                    'node_id': node.id,
                    'duration_seconds': duration,
                    'iterations': result_data.get('iterations', 0)
                }
            )

            return result

        except NotImplementedError as exc:
            # Expected error for Week 8 (Celery task not implemented yet)
            error_msg = f"Loop execution not available: {str(exc)}"
            logger.warning(error_msg, extra={'node_id': node.id})

            result = NodeExecutionResult(
                success=False,
                output=None,
                error=error_msg,
                mode=NodeExecutionMode.SYNC,
                duration_seconds=time.time() - start_time
            )
            self._update_step_result(step_result, result)
            return result

        except Exception as exc:
            error_msg = f"Failed to execute loop node: {str(exc)}"
            logger.error(
                error_msg,
                extra={'node_id': node.id},
                exc_info=True
            )

            result = NodeExecutionResult(
                success=False,
                output=None,
                error=error_msg,
                mode=NodeExecutionMode.SYNC,
                duration_seconds=time.time() - start_time
            )
            self._update_step_result(step_result, result)
            return result

    def _execute_count_loop(
        self,
        loop_config: Any,
        loop_node_id: str,
        context: Dict[str, Any],
        execution: WorkflowExecution,
        max_iterations: int
    ) -> Dict[str, Any]:
        """
        Execute count loop: for i in range(N).

        Args:
            loop_config: LoopConfig Pydantic schema
            loop_node_id: Node ID to execute in loop
            context: Execution context
            execution: WorkflowExecution instance
            max_iterations: Maximum iterations allowed

        Returns:
            Dict with iteration results
        """
        if not loop_config.count:
            raise ValueError("count is required for mode='count'")

        count = min(loop_config.count, max_iterations)
        results = []

        for i in range(count):
            # Update context with loop variables
            loop_context = context.copy()
            loop_context['loop'] = {
                'index': i,
                'first': i == 0,
                'last': i == count - 1,
            }

            # Execute loop node (placeholder - will use Celery in Week 9)
            from apps.templates.tasks import execute_workflow_node
            result = execute_workflow_node(str(execution.id), loop_node_id, loop_context)
            results.append({'iteration': i, 'result': result})

        return {
            'mode': 'count',
            'iterations': count,
            'results': results
        }

    def _execute_while_loop(
        self,
        loop_config: Any,
        loop_node_id: str,
        context: Dict[str, Any],
        execution: WorkflowExecution,
        max_iterations: int
    ) -> Dict[str, Any]:
        """
        Execute while loop: while condition_is_true.

        Args:
            loop_config: LoopConfig Pydantic schema
            loop_node_id: Node ID to execute in loop
            context: Execution context
            execution: WorkflowExecution instance
            max_iterations: Maximum iterations allowed

        Returns:
            Dict with iteration results
        """
        if not loop_config.condition:
            raise ValueError("condition is required for mode='while'")

        condition_expr = loop_config.condition
        results = []
        iteration = 0

        while iteration < max_iterations:
            # Update context with loop variables
            loop_context = context.copy()
            loop_context['loop'] = {
                'index': iteration,
                'first': iteration == 0,
            }

            # Evaluate condition in sandbox
            try:
                template = self.env.from_string(condition_expr)
                rendered = template.render(loop_context)
                should_continue = self._to_bool(rendered)
            except Exception as exc:
                logger.error(f"Failed to evaluate while condition: {exc}", exc_info=True)
                break

            if not should_continue:
                break

            # Execute loop node (placeholder - will use Celery in Week 9)
            from apps.templates.tasks import execute_workflow_node
            result = execute_workflow_node(str(execution.id), loop_node_id, loop_context)
            results.append({'iteration': iteration, 'result': result})

            # Update context with result for next iteration
            context.update(result)
            iteration += 1

        return {
            'mode': 'while',
            'iterations': iteration,
            'results': results,
            'max_iterations_reached': iteration >= max_iterations
        }

    def _execute_foreach_loop(
        self,
        loop_config: Any,
        loop_node_id: str,
        context: Dict[str, Any],
        execution: WorkflowExecution,
        max_iterations: int
    ) -> Dict[str, Any]:
        """
        Execute foreach loop: for item in collection.

        Args:
            loop_config: LoopConfig Pydantic schema
            loop_node_id: Node ID to execute in loop
            context: Execution context
            execution: WorkflowExecution instance
            max_iterations: Maximum iterations allowed

        Returns:
            Dict with iteration results
        """
        if not loop_config.items:
            raise ValueError("items is required for mode='foreach'")

        # Get items collection from context using dot notation
        items = self._resolve_path(context, loop_config.items)

        if not isinstance(items, (list, tuple)):
            raise ValueError(f"items must be a list or tuple, got {type(items).__name__}")

        # Limit iterations
        items_limited = items[:max_iterations]
        results = []

        for i, item in enumerate(items_limited):
            # Update context with loop variables
            loop_context = context.copy()
            loop_context['loop'] = {
                'index': i,
                'first': i == 0,
                'last': i == len(items_limited) - 1,
            }
            loop_context['item'] = item

            # Execute loop node (placeholder - will use Celery in Week 9)
            from apps.templates.tasks import execute_workflow_node
            result = execute_workflow_node(str(execution.id), loop_node_id, loop_context)
            results.append({'iteration': i, 'item': item, 'result': result})

        return {
            'mode': 'foreach',
            'iterations': len(items_limited),
            'total_items': len(items),
            'truncated': len(items) > max_iterations,
            'results': results
        }

    def _resolve_path(self, context: Dict[str, Any], path: str) -> Any:
        """
        Resolve dot notation path in context.

        Args:
            context: Execution context
            path: Dot notation path (e.g., 'data.users')

        Returns:
            Resolved value

        Raises:
            KeyError: If path not found
        """
        keys = path.split('.')
        value = context

        for key in keys:
            if isinstance(value, dict):
                value = value[key]
            else:
                raise KeyError(f"Cannot resolve path '{path}': '{key}' not found")

        return value

    def _to_bool(self, value: Any) -> bool:
        """
        Convert rendered value to boolean.

        Args:
            value: Rendered expression value

        Returns:
            bool: Converted boolean value
        """
        # Direct boolean
        if isinstance(value, bool):
            return value

        # String conversion (case-insensitive)
        if isinstance(value, str):
            value_lower = value.strip().lower()
            if value_lower in ('true', 'yes', '1'):
                return True
            elif value_lower in ('false', 'no', '0', '', 'none'):
                return False
            else:
                # Non-empty string -> True (Pythonic)
                return bool(value_lower)

        # Integer conversion
        if isinstance(value, int):
            return value != 0

        # None -> False
        if value is None:
            return False

        # Collections (list, dict, etc.)
        # Empty -> False, non-empty -> True
        return bool(value)

File: go-services/worker/internal/workflow/handlers/loop.go
package handlers

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"go.uber.org/zap"

	wfcontext "github.com/commandcenter1c/commandcenter/worker/internal/workflow/context"
	"github.com/commandcenter1c/commandcenter/worker/internal/workflow/executor"
	"github.com/commandcenter1c/commandcenter/worker/internal/workflow/models"
)

// MaxIterationsHardLimit is the safety limit for loop iterations.
const MaxIterationsHardLimit = 10000

// LoopHandler executes loop nodes with count, while, and foreach modes.
//
// Loop modes:
//   - count: Execute body N times (for i in range(N))
//   - while: Execute body while condition is true
//   - foreach: Execute body for each item in collection
//
// Context updates for each iteration:
//   - loop.index: Current iteration index (0-based)
//   - loop.first: True if first iteration
//   - loop.last: True if last iteration (count/foreach only)
//   - item: Current item (foreach mode only)
//
// Safety:
//   - Hard limit of 10000 iterations
//   - Respects max_iterations from config
//   - Checks context cancellation between iterations
type LoopHandler struct {
	registry       *HandlerRegistry
	templateEngine TemplateRenderer
	logger         *zap.Logger
}

// NewLoopHandler creates a new loop handler.
func NewLoopHandler(deps *HandlerDependencies, registry *HandlerRegistry) *LoopHandler {
	logger := deps.Logger
	if logger == nil {
		logger = zap.NewNop()
	}
	return &LoopHandler{
		registry:       registry,
		templateEngine: deps.TemplateEngine,
		logger:         logger.Named("loop_handler"),
	}
}

// SupportedTypes returns the node types this handler can process.
func (h *LoopHandler) SupportedTypes() []models.NodeType {
	return []models.NodeType{models.NodeTypeLoop}
}

// IterationResult holds the result of a single loop iteration.
type IterationResult struct {
	Iteration int
	Item      interface{}
	Result    interface{}
	Error     error
}

// HandleNode executes a loop based on its configuration.
func (h *LoopHandler) HandleNode(
	ctx context.Context,
	node *models.Node,
	execCtx *wfcontext.ExecutionContext,
) (*executor.NodeResult, error) {
	startTime := time.Now()

	h.logger.Debug("Executing loop node",
		zap.String("node_id", node.ID),
		zap.String("node_name", node.Name),
		zap.String("execution_id", execCtx.ExecutionID()))

	// Get loop config
	config := node.LoopConfig
	if config == nil {
		h.logger.Error("Loop node missing config",
			zap.String("node_id", node.ID))
		return &executor.NodeResult{
			NodeID:      node.ID,
			Status:      executor.NodeStatusFailed,
			Error:       fmt.Errorf("loop node %s missing loop_config", node.ID),
			StartedAt:   startTime,
			CompletedAt: time.Now(),
			Duration:    time.Since(startTime),
		}, nil
	}

	// Calculate max iterations
	maxIterations := config.MaxIterations
	if maxIterations <= 0 || maxIterations > MaxIterationsHardLimit {
		maxIterations = MaxIterationsHardLimit
	}

	h.logger.Debug("Loop config",
		zap.String("node_id", node.ID),
		zap.String("mode", string(config.Mode)),
		zap.String("body_node", config.BodyNode),
		zap.Int("max_iterations", maxIterations))

	// Execute loop based on mode
	var result *LoopResult
	var err error

	switch config.Mode {
	case models.LoopModeCount:
		result, err = h.executeCountLoop(ctx, node, config, execCtx, maxIterations)
	case models.LoopModeWhile:
		result, err = h.executeWhileLoop(ctx, node, config, execCtx, maxIterations)
	case models.LoopModeForeach:
		result, err = h.executeForeachLoop(ctx, node, config, execCtx, maxIterations)
	default:
		err = fmt.Errorf("unknown loop mode: %s", config.Mode)
	}

	if err != nil {
		h.logger.Error("Loop execution failed",
			zap.String("node_id", node.ID),
			zap.Error(err))
		return &executor.NodeResult{
			NodeID:      node.ID,
			Status:      executor.NodeStatusFailed,
			Error:       err,
			StartedAt:   startTime,
			CompletedAt: time.Now(),
			Duration:    time.Since(startTime),
		}, nil
	}

	h.logger.Debug("Loop execution completed",
		zap.String("node_id", node.ID),
		zap.Int("iterations", result.Iterations),
		zap.Duration("duration", time.Since(startTime)))

	return &executor.NodeResult{
		NodeID: node.ID,
		Status: executor.NodeStatusCompleted,
		Output: map[string]interface{}{
			"mode":                   string(result.Mode),
			"iterations":             result.Iterations,
			"results":                result.Results,
			"max_iterations_reached": result.MaxIterationsReached,
			"total_items":            result.TotalItems,
			"truncated":              result.Truncated,
		},
		StartedAt:   startTime,
		CompletedAt: time.Now(),
		Duration:    time.Since(startTime),
	}, nil
}

// LoopResult holds the aggregated results of loop execution.
type LoopResult struct {
	Mode                 models.LoopMode
	Iterations           int
	Results              []interface{}
	MaxIterationsReached bool
	TotalItems           int
	Truncated            bool
}

// executeCountLoop executes a fixed number of iterations.
func (h *LoopHandler) executeCountLoop(
	ctx context.Context,
	node *models.Node,
	config *models.LoopNodeConfig,
	execCtx *wfcontext.ExecutionContext,
	maxIterations int,
) (*LoopResult, error) {
	if config.Count <= 0 {
		return nil, fmt.Errorf("count must be positive for count loop mode")
	}

	count := config.Count
	if count > maxIterations {
		count = maxIterations
	}

	result := &LoopResult{
		Mode:    models.LoopModeCount,
		Results: make([]interface{}, 0, count),
	}

	for i := 0; i < count; i++ {
		// Check for cancellation
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		default:
		}

		// Execute iteration
		iterResult, err := h.executeIteration(ctx, node, config, execCtx, i, count, nil)
		if err != nil {
			return nil, fmt.Errorf("loop iteration %d failed: %w", i, err)
		}

		result.Results = append(result.Results, map[string]interface{}{
			"iteration": i,
			"result":    iterResult,
		})
		result.Iterations++
	}

	result.MaxIterationsReached = config.Count > maxIterations

	return result, nil
}

// executeWhileLoop executes while condition is true.
func (h *LoopHandler) executeWhileLoop(
	ctx context.Context,
	node *models.Node,
	config *models.LoopNodeConfig,
	execCtx *wfcontext.ExecutionContext,
	maxIterations int,
) (*LoopResult, error) {
	if config.Condition == "" {
		return nil, fmt.Errorf("condition is required for while loop mode")
	}

	result := &LoopResult{
		Mode:    models.LoopModeWhile,
		Results: make([]interface{}, 0),
	}

	// Working context that accumulates results
	workingCtx := execCtx

	for iteration := 0; iteration < maxIterations; iteration++ {
		// Check for cancellation
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		default:
		}

		// Evaluate condition
		shouldContinue, err := h.evaluateCondition(ctx, config.Condition, workingCtx, iteration)
		if err != nil {
			h.logger.Error("Failed to evaluate while condition",
				zap.String("node_id", node.ID),
				zap.Int("iteration", iteration),
				zap.Error(err))
			break
		}

		if !shouldContinue {
			break
		}

		// Execute iteration
		iterResult, err := h.executeIteration(ctx, node, config, workingCtx, iteration, -1, nil)
		if err != nil {
			return nil, fmt.Errorf("loop iteration %d failed: %w", iteration, err)
		}

		result.Results = append(result.Results, map[string]interface{}{
			"iteration": iteration,
			"result":    iterResult,
		})
		result.Iterations++

		// Update working context with iteration result
		if iterResultMap, ok := iterResult.(map[string]interface{}); ok {
			workingCtx = workingCtx.Merge(iterResultMap)
		}
	}

	result.MaxIterationsReached = result.Iterations >= maxIterations

	return result, nil
}

// executeForeachLoop executes for each item in collection.
func (h *LoopHandler) executeForeachLoop(
	ctx context.Context,
	node *models.Node,
	config *models.LoopNodeConfig,
	execCtx *wfcontext.ExecutionContext,
	maxIterations int,
) (*LoopResult, error) {
	if config.Items == "" {
		return nil, fmt.Errorf("items is required for foreach loop mode")
	}

	// Resolve items collection from context
	items, err := h.resolveItems(ctx, config.Items, execCtx)
	if err != nil {
		return nil, fmt.Errorf("failed to resolve items: %w", err)
	}

	result := &LoopResult{
		Mode:       models.LoopModeForeach,
		Results:    make([]interface{}, 0),
		TotalItems: len(items),
	}

	// Apply max iterations limit
	itemsToProcess := items
	if len(items) > maxIterations {
		itemsToProcess = items[:maxIterations]
		result.Truncated = true
	}

	for i, item := range itemsToProcess {
		// Check for cancellation
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		default:
		}

		// Execute iteration with current item
		iterResult, err := h.executeIteration(ctx, node, config, execCtx, i, len(itemsToProcess), item)
		if err != nil {
			return nil, fmt.Errorf("loop iteration %d failed: %w", i, err)
		}

		result.Results = append(result.Results, map[string]interface{}{
			"iteration": i,
			"item":      item,
			"result":    iterResult,
		})
		result.Iterations++
	}

	return result, nil
}

// executeIteration executes a single loop iteration.
func (h *LoopHandler) executeIteration(
	ctx context.Context,
	node *models.Node,
	config *models.LoopNodeConfig,
	execCtx *wfcontext.ExecutionContext,
	index int,
	totalCount int,
	item interface{},
) (interface{}, error) {
	// Placeholder implementation: iteration context is only used in ExecuteIterationWithDAG.

	// In production, we would execute the body node here.
	// For now, return a placeholder result.
	h.logger.Debug("Executing loop iteration",
		zap.String("node_id", node.ID),
		zap.String("body_node", config.BodyNode),
		zap.Int("iteration", index))

	// Placeholder result - in production, this would call the handler for body_node
	return map[string]interface{}{
		"iteration_executed": true,
		"body_node":          config.BodyNode,
		"index":              index,
	}, nil
}

// ExecuteIterationWithDAG executes a single loop iteration with access to the full DAG.
// This is the production version that should be used when DAG is available.
func (h *LoopHandler) ExecuteIterationWithDAG(
	ctx context.Context,
	dag *models.DAG,
	config *models.LoopNodeConfig,
	execCtx *wfcontext.ExecutionContext,
	index int,
	totalCount int,
	item interface{},
) (interface{}, error) {
	// Create iteration context with loop variables
	iterCtx := execCtx.PushScope()

	// Set loop variables
	loopVars := map[string]interface{}{
		"index": index,
		"first": index == 0,
	}
	if totalCount >= 0 {
		loopVars["last"] = index == totalCount-1
	}
	iterCtx = iterCtx.SetScoped("loop", loopVars)
	iterCtx = iterCtx.SetScoped("loop_index", index)
	iterCtx = iterCtx.SetScoped("loop_first", index == 0)
	if totalCount >= 0 {
		iterCtx = iterCtx.SetScoped("loop_last", index == totalCount-1)
	}

	// Set item for foreach mode
	if item != nil {
		loopVar := config.LoopVar
		if loopVar == "" {
			loopVar = "item"
		}
		iterCtx = iterCtx.SetScoped(loopVar, item)
		iterCtx = iterCtx.SetScoped("item", item)
	}

	// Get body node from DAG
	bodyNode := dag.GetNode(config.BodyNode)
	if bodyNode == nil {
		return nil, fmt.Errorf("body node %s not found in DAG", config.BodyNode)
	}

	// Get handler for body node
	handler, ok := h.registry.Get(bodyNode.Type)
	if !ok {
		return nil, fmt.Errorf("no handler for node type: %s", bodyNode.Type)
	}

	// Execute body node
	result, err := handler.HandleNode(ctx, bodyNode, iterCtx)
	if err != nil {
		return nil, err
	}

	return result.Output, nil
}

// evaluateCondition evaluates a while loop condition.
func (h *LoopHandler) evaluateCondition(
	ctx context.Context,
	condition string,
	execCtx *wfcontext.ExecutionContext,
	iteration int,
) (bool, error) {
	// Add iteration to context for condition evaluation
	evalCtx := execCtx.SetScoped("loop_index", iteration)

	if h.templateEngine != nil {
		renderCtx := evalCtx.ToMap()
		rendered, err := h.templateEngine.Render(ctx, condition, renderCtx)
		if err != nil {
			return false, err
		}
		return isTruthy(rendered), nil
	}

	// Simple evaluation without template engine
	return isTruthy(condition), nil
}

// resolveItems resolves the items expression to a slice.
func (h *LoopHandler) resolveItems(
	ctx context.Context,
	itemsExpr string,
	execCtx *wfcontext.ExecutionContext,
) ([]interface{}, error) {
	// Try to render with template engine
	if h.templateEngine != nil {
		renderCtx := execCtx.ToMap()
		rendered, err := h.templateEngine.Render(ctx, itemsExpr, renderCtx)
		if err != nil {
			return nil, fmt.Errorf("failed to render items expression: %w", err)
		}

		// Try to parse as JSON array
		var items []interface{}
		if err := json.Unmarshal([]byte(rendered), &items); err == nil {
			return items, nil
		}

		// If not JSON, try as simple string
		return []interface{}{rendered}, nil
	}

	// Try to resolve from context using dot notation
	return h.resolvePath(execCtx, itemsExpr)
}

// resolvePath resolves a dot notation path to a slice.
func (h *LoopHandler) resolvePath(
	execCtx *wfcontext.ExecutionContext,
	path string,
) ([]interface{}, error) {
	// Clean path (remove template markers if present)
	path = strings.TrimSpace(path)
	path = strings.TrimPrefix(path, "{{")
	path = strings.TrimSuffix(path, "}}")
	path = strings.TrimSpace(path)

	// Get value from context
	val, ok := execCtx.Get(path)
	if !ok {
		return nil, fmt.Errorf("path '%s' not found in context", path)
	}

	// Convert to slice
	switch v := val.(type) {
	case []interface{}:
		return v, nil
	case []string:
		result := make([]interface{}, len(v))
		for i, s := range v {
			result[i] = s
		}
		return result, nil
	case []int:
		result := make([]interface{}, len(v))
		for i, n := range v {
			result[i] = n
		}
		return result, nil
	case []map[string]interface{}:
		result := make([]interface{}, len(v))
		for i, m := range v {
			result[i] = m
		}
		return result, nil
	default:
		return nil, fmt.Errorf("path '%s' is not a list (got %T)", path, val)
	}
}
